cpp/src/lists/drop_list_duplicates.cu

```diff
@@ -0,0 +1,288 @@
+    [num_entries, offsets_ptr = entries_list_offsets.begin<size_type>()] __device__(size_type i)
+      -> bool { return i == 0 || i == num_entries || offsets_ptr[i] != offsets_ptr[i - 1]; });
+
+  // Generate a prefix sum of number of empty lists, storing inplace to the original lists
+  // offsets
+  // If the original list offsets is { 0, 0, 5, 5, 6, 6 } (there are 2 empty lists),
+  // and new_offsets = { 0, 4, 6 },
+  // then output = { 0, 1, 1, 2, 2, 3}
+  auto const iter_trans_begin = cudf::detail::make_counting_transform_iterator(
+    0, [offsets = original_offsets.begin<size_type>()] __device__(size_type i) -> size_type {
+      return (i > 0 && offsets[i] == offsets[i - 1]) ? 1 : 0;
+    });
+  thrust::inclusive_scan(rmm::exec_policy(stream),
+                         iter_trans_begin,
+                         iter_trans_begin + original_offsets.size(),
+                         original_offsets.begin<size_type>());
+
+  // Generate the final list offsets
+  // If the original list offsets are { 0, 0, 5, 5, 6, 6 }, the new offsets are { 0, 4, 6 },
+  //  and the prefix sums of empty lists are { 0, 1, 1, 2, 2, 3 },
+  //  then output = { 0, 0, 4, 4, 5, 5 }
+  thrust::transform(
+    rmm::exec_policy(stream),
+    thrust::make_counting_iterator<size_type>(0),
+    thrust::make_counting_iterator<size_type>(original_offsets.size()),
+    original_offsets.begin<size_type>(),
+    [prefix_sum_empty_lists = original_offsets.begin<size_type>(),
+     offsets = new_offsets->view().begin<size_type>()] __device__(size_type i) -> size_type {
+      return offsets[i - prefix_sum_empty_lists[i]];
+    });
+}
+/**
+ * @copydoc cudf::lists::drop_list_duplicates
+ *
+ * @param stream CUDA stream used for device memory operations and kernel launches
+ */
+std::unique_ptr<column> drop_list_duplicates(lists_column_view const& lists_column,
+                                             null_equality nulls_equal,
+                                             rmm::cuda_stream_view stream,
+                                             rmm::mr::device_memory_resource* mr)
+{
+  if (lists_column.is_empty()) return cudf::empty_like(lists_column.parent());
+  if (cudf::is_nested(lists_column.child().type()))
+    CUDF_FAIL("Nested types are not supported in drop_list_duplicates.");
+
+  // Call segmented sort on the list elements
+  auto const sorted_lists = lists::sort_lists(lists_column, order::ASCENDING, null_order::AFTER);
+
+  // Flatten all entries (depth = 1) of the lists column
+  auto const all_lists_entries = lists_column_view(sorted_lists->view()).get_sliced_child(stream);
+
+  // Generate a 0-based offset column
+  auto lists_offsets = detail::generate_clean_offsets(lists_column, stream, mr);
+
+  // Generate a mapping from list entries to offsets of the lists containing those entries
+  auto const entries_list_offsets = detail::generate_entry_list_offsets(
+    all_lists_entries.size(), lists_offsets->view(), stream, mr);
```

Comment:
I think `entries_list_offsets` is a temporary column and so should not use `mr` to build it.

Reply:
Why not? Do you have a better suggestion? We want to have fast memory allocation. If not using mr, then we may require memory from the system, which is much slower.

Reply:
The passed in `mr` is used for allocating any memory that would be returned and managed by the caller. We should not to use that resource for allocating temporary memory. Just use a default memory resource (`rmm::mr::get_current_device_resource()`) instance for allocating temporary columns and such instead.
Here is the section on temporary memory in the developer guide.
https://github.com/rapidsai/cudf/blob/branch-0.19/cpp/docs/DEVELOPER_GUIDE.md#temporary-memory

Here is an example creating a temporary column in `sort.cu`
https://github.com/rapidsai/cudf/blob/53929eb7ce03954a7b107a2e08639da3447b0ecf/cpp/src/sort/sort.cu#L48-L56
The `sorted_order` column is not returned and the default memory resource is used to allocate it. The result of the `gather` is returned to the caller and so is passed the `mr` for it to allocate the returned column.

Reply:
Thanks so much for this. I have read the developer guide multiple times, but somehow I can't remember the right points. I must definitely read it again now.